src/serac/numerics/functional/functional_qoi.inl

```diff
@@ -58,64 +58,81 @@ struct QoIElementRestriction : public mfem::Operator {
 /**
  * @brief a partial template specialization of Functional with test == double, implying "quantity of interest"
  */
-template <typename trial, ExecutionSpace exec>
-class Functional<double(trial), exec> : public mfem::Operator {
+template <typename... trials, ExecutionSpace exec>
+class Functional<double(trials...), exec> {
   using test = QOI;
+  static constexpr tuple<trials...> trial_spaces{};
+  static constexpr uint32_t         num_trial_spaces = sizeof...(trials);
+
+  class Gradient;
+
+  template <typename... T>
+  struct operator_paren_return {
+    using type = typename std::conditional<
+      (std::is_same_v<T, dual_vector> + ...) == 1,  // if the there is a dual number in the pack
```

Comment:
I really like the readability of this metaprogramming. Thank you for the comments!